library/std/src/io/stdio.rs

```diff
@@ -1017,6 +1018,145 @@ where
     }
 }
 
+mod terminal_sealed {
+    #[unstable(feature = "sealed", issue = "none")]
+    pub trait Sealed {}
+}
+
+/// Trait to determine if a descriptor/handle refers to a terminal/tty.
+#[unstable(feature = "is_terminal", issue = "98070")]
+pub trait IsTerminal: terminal_sealed::Sealed {
+    /// Returns true if the descriptor/handle refers to a terminal/tty.
+    ///
+    /// Returns an error if any errors occurred when attempting to make this determination, other
+    /// than those expected as part of making this determination. For instance, on some platforms
+    /// checking for a terminal may involve making a system call that returns an error on a
+    /// non-terminal, so that error will get translated to `Ok(false)`, but any unexpected error
+    /// will still get returned as an error.
```

Comment:
`isatty` doesn't do any actual I/O. It does use a system call, but there's no external reason for the system call to ever fail because it doesn't do anything. I'm not advocating for all APIs to panic on `EBADF` at this point, but it does seem like panicking on `EBADF` would be reasonable for `is_terminal` specifically. That would let it return a convenient `bool` instead of `io::Result<bool>`, and avoid the need for this subtle documentation comment, without loss of realistic functionality.

Reply:
I had been thinking that we may want to reserve the ability for platforms to return an error here. But I think you're right that people won't actually want that in practice. Together with the EINVAL point you raised above, and other kernel bugs (some versions of Linux returned EFAULT in some cases as recently as 2021), I think it'd be reasonable to just treat any error as `false`.